File: 916-word-subsets/916-word-subsets.java
class Solution {
    public List<String> wordSubsets(String[] words1, String[] words2) {
        int[] maxCharCount = new int[26];
        for(String word : words2){
            int[] charCount = new int[26];
            for(int i = 0; i < word.length(); i++){
                int j = word.charAt(i) - 'a';
                charCount[j]++;
                maxCharCount[j] = Math.max(maxCharCount[j], charCount[j]);
            }
        }
        
        List<String> ans = new ArrayList<>();
        for(String word : words1){
            int[] charCount = new int[26];
            int i = 0;
            for(; i < word.length(); i++){
                int j = word.charAt(i) - 'a';
                charCount[j]++;
            }
            
            for(i = 0; i < 26; i++){
                if(maxCharCount[i] > charCount[i]){
                    break;
                }
            }
            
            if(i != 26){
                continue;
            }
            
            ans.add(word);
        }
        
        return ans;
    }
}

File: 916-word-subsets/916-word-subsets.py
class Solution:
    def wordSubsets(self, words1: List[str], words2: List[str]) -> List[str]:
        max_char_count = [0] * 26
        for word in words2:
            for c, count in Counter(word).items():
                i = ord(c) - ord('a')
                max_char_count[i] = max(max_char_count[i], count)
        
        ans = []
        for word in words1:
            counts = Counter(word)
            for i in range(26):
                if counts[chr(ord('a') + i)] < max_char_count[i]:
                    break
            else:
                ans.append(word)
        return ans